### DISCUSSION TITLE
Whole planet build

### DISCUSSION THREAD

**Discussion by rpmcoding** [systemed/tilemaker#434] [Q&A] [answered]
[systemed](https://github.com/systemed/tilemaker/commits?author=systemed) you mention that if you want a full planet build (as we do) you should consider an alternative solution.

What other solutions would you suggest?

Thanks so much.

**Comment by systemed** [comment, marked as answer]
You can do a whole planet build with tilemaker now, if you have a lot of RAM and SSD. #315 has some comments. Note that it's best to use the bounding box in that example (-60° to 75° latitude), i.e. excluding the polar regions - tilemaker doesn't cope very well with the polar regions at present.

You could look at https://github.com/onthegomap/planetiler which has lower memory and disk requirements - if you just want to create an OpenMapTiles-compatible planet then this might be a good way to go.

**Reply by rpmcoding** [reply]
@systemed thanks, really appreciate that.

**Comment by sujithkris** [comment]
Whether anyone tried the whole planet with tilemaker and benchmarked the minimum system requirement?
I tried in an AWS instance with 64CPU/128GB/1900GBSSD. The full planet conversion is **crashing the process**. 
Tried with `coastline/water_ploygons` and `natural earth urban areas`.

Any help?

**Comment by systemed** [comment]
Have you excluded the polar regions - i.e. by specifying a bounding box restricted to -60° to 75° latitude? That will make a big difference.

**Reply by sujithkris** [reply]
> Have you excluded the polar regions - i.e. by specifying a bounding box restricted to -60° to 75° latitude? That will make a big difference.

oh. I missed that I think. Will exclude the polar and give it a try. I am using `--store` option as well.

Also, I am generating vector tiles with max_zoom as 19, ( `--output` is a directory) rather than `.mbtiles` file. Do you think `--compact` can also help? I do not want to lose the city/town/road data, same time, I am least bothered on national boundary accuracies.

**UPDATE**
Tried
```
tilemaker --input /mount1/osmpbf/planet-latest.osm.pbf --output /mount1/tiles --bbox -180,-60,180,75 --compact --store /mount2/store --config /mount1/configs/zoom_19_config.json
```
I could see the process getting killed even without a spike in CPU and Memory !!
 Looks like higher zoom levels + shapes files not succeeding.


With default config (zoom level 14) the process is proceeding with shapes and then planet pbf. ( works fine)

**Reply by systemed** [reply]
You don't need to generate vector tiles at z19 (and tilemaker won't work for you if you try!). Just generate them at z14 and then use the data in those tiles to render all zoom levels below. It's not the same as raster tiles where you would generate z19 tiles.

**Reply by sujithkris** [reply]
> You don't need to generate vector tiles at z19 (and tilemaker won't work for you if you try!). Just generate them at z14 and then use the data in those tiles to render all zoom levels below. It's not the same as raster tiles where you would generate z19 tiles.

Sure. I thought of trying by defining `FAT_TILE_INDEX` to generate the tiles above z16. But let me limit it to z14 and test it.
@systemed  - Whether you observed any issues when using `coastline` and `landcover`?
For me, the tile generation is not starting when using `coastline` and `landcover`. The process just killed without much CPU/memory spike after the `store` file creation

However, tile generation working fine with the default config (z14) without `coastline` and `landcover`

**Reply by systemed** [reply]
Essentially the reason to exclude the polar regions is because the coastline/landcover creates a _lot_ of tiles here. Because of the way the Mercator projection works, tilemaker ends up creating vast numbers of objects for the coastline/landcover at the poles, using about 27GB extra RAM.

But if you don't have any objects in these areas, or you exclude these areas automatically, the objects won't get created.

https://github.com/systemed/tilemaker/issues/293#issuecomment-904828329 refers. #314 and #323 were attempts to fix this but both had problems.

**Reply by sujithkris** [reply]
> Essentially the reason to exclude the polar regions is because the coastline/landcover creates a _lot_ of tiles here. Because of the way the Mercator projection works, tilemaker ends up creating vast numbers of objects for the coastline/landcover at the poles, using about 27GB extra RAM.
> 
> But if you don't have any objects in these areas, or you exclude these areas automatically, the objects won't get created.
> 
> [#293 (comment)](https://github.com/systemed/tilemaker/issues/293#issuecomment-904828329) refers. #314 and #323 were attempts to fix this but both had problems.

Alright. Thanks for the clarification. 
Newbie question - For the tile request for z15-z19, I can return the same z14 tiles from my tile server? My idea is to place an indoor map of the building on top of this planet map, hence zoom levels of 15+ are widely used.  Primarily trying to migrate from Mapbox which has planet tiles till z22.

**Reply by systemed** [reply]
Yes, you can just use the z14 tiles. The vector data in these can be magnified for higher zoom levels.

Because tilemaker pre-generates vector tiles (rather than serving them on demand), generating everything up to z19 would be massively slow and memory-hungry.

If you're planning to go up to high zoom levels, you might want to set the `high_resolution` parameter in your JSON config (see #384). This doesn't affect RAM consumption and only makes the tiles slightly bigger, but will result in greater accuracy when you're zoomed in.

**Reply by sujithkris** [reply]
> Yes, you can just use the z14 tiles. The vector data in these can be magnified for higher zoom levels.
> 
> Because tilemaker pre-generates vector tiles (rather than serving them on demand), generating everything up to z19 would be massively slow and memory-hungry.
> 
> If you're planning to go up to high zoom levels, you might want to set the `high_resolution` parameter in your JSON config (see #384). This doesn't affect RAM consumption and only makes the tiles slightly bigger, but will result in greater accuracy when you're zoomed in.

Highly insightful. Thanks

I am generating tiles using tilemaker, pushing tiles to CDN, and then serving from CDN


So we need a system with min 150GB memory to accommodate `coastline` and `landcover/ne_urban_areas`, right ? For me process crashed in the system of 128GB. but successfully generated 0-14 tiles with the default config without the above two shapes.

**Reply by systemed** [reply]
Interesting. I've run the planet successfully with 144GB (a few months ago, and obviously the data is always getting bigger). I have a long-term idea about reducing memory usage but that's going to take some heavy lifting; I wonder if there's anything more that could be done in the short term to reduce the coastline data overhead.

One thing you can do to save on memory is run the coastline/landcover and OSM-derived data separately. In other words, first run tilemaker without specifying a .pbf, and with a full bounding box (within the limits of the Mercator projection, so roughly -85° to 85° latitude). This will generate an .mbtiles containing the shapefile-derived data only. (It's worthwhile saving a copy of this somewhere.)

Then edit out the shapefile layers from the .json file, and rerun tilemaker with planet-latest.osm.pbf, this time using the `--merge` option with your newly created .mbtiles. It will overlay the OSM-derived data onto the already-created tiles with coastline etc.

**Reply by systemed** [reply]
#479 is a revival of #323 which I hope may be more efficient. Still testing.

**Comment by sujithkris** [comment]
@systemed  : Not getting the (state/country boundaries) for planet osm pbf. Is this because I am not including the coastline/landcover shapes? 

Tilemaker
<img width="1396" alt="Screenshot 2023-03-27 at 1 28 33 PM" src="https://user-images.githubusercontent.com/2143097/227878020-fb7dac41-7159-496f-984e-06770b7a811f.png">

Mapbox
<img width="1329" alt="Screenshot 2023-03-27 at 1 28 22 PM" src="https://user-images.githubusercontent.com/2143097/227878040-179a8fd7-9a2d-4452-a4dc-6a87e51d8006.png">

**Reply by systemed** [reply]
tilemaker's OMT-compatible config does write boundaries (https://github.com/systemed/tilemaker/blob/master/resources/process-openmaptiles.lua#L249-L285) so they should be in the vector tiles. Maybe a stylesheet issue?

**Reply by sujithkris** [reply]
@systemed  - I directly applied the styling from Maptiler, which again failed to draw the boundaries with tilemaker tiles. I am using tilemaker master branch.
My [tilemaker-config.json](https://github.com/systemed/tilemaker/files/11156941/tilemaker-config.json.txt)

Also, very few `ocean` tiles are available (yes, I am using coastline shape also). However, almost all `lakes` are present.
<img width="1046" alt="Screenshot 2023-04-05 at 2 35 54 PM" src="https://user-images.githubusercontent.com/2143097/230035244-49fc5982-3f8c-4f5f-bfbd-d5c24951183e.png">
This screenshot is with `zoom 6.3` and `center {lng: 79.34599067516137, lat: 8.080629437792709}`

Do you recommend any stylesheet to check whether features are generated properly by tilemaker? ( to isolate any stylesheet issues)

**Reply by systemed** [reply]
The best way to see what's in a vector tile is to use an Inspect plugin: https://github.com/acalcutt/maplibre-gl-inspect

It's also sometimes helpful to use [vt2geojson](https://github.com/mapbox/vt2geojson) to fetch and decompile a tile into GeoJSON, which you can then inspect in QGIS or whatever your favoured editor is!

**Reply by sujithkris** [reply]
inspect helped!
```
"filter": [
        "all",
        [
          "in",
          "admin_level",
          3,
          4
        ],
        [
          "==",
          "maritime",
          0
        ]
      ],
```
I could see tilemaker generated tiles don't have `maritime : 0` and my boundary styles not applied.
I took the styles as is from https://github.com/openmaptiles/openmaptiles/blob/master/layers/boundary/style.json

**Comment by sujithkris** [comment]
<img width="1522" alt="Screenshot 2023-03-29 at 8 04 05 AM" src="https://user-images.githubusercontent.com/2143097/228411811-b6c7e86e-1863-402f-a872-357942cc7992.png">
@systemed: Do we have a log file to keep track of progress? For me, the process looks like in _idle/sleep_. 
Generated store files (last modified is 30hrs before) but yet to generate tiles.


**System**
_256.0 GiB | 16 vCPUs | 950 GB_

**Process**
`tilemaker --input /data/osm/planet-latest.osm.pbf --output /data/tiles --store /data/store --config 16_config.json --bbox -180,-60,180,75 --compact`


I am running for minzoom 15 and maxzoom 16.

```
"settings": {
		"minzoom": 15,
		"maxzoom": 16,
		"basezoom": 16,
		"include_ids": false,
		"combine_below": 14,
		"name": "Tilemaker to OpenMapTiles schema",
		"version": "3.0",
		"description": "Tile config based on OpenMapTiles schema",
		"compress": "gzip",
		"filemetadata": {
			"tilejson": "2.0.0", 
			"scheme": "xyz", 
			"type": "baselayer", 
			"format": "pbf", 
            "tiles": ["https://example.com/liechtenstein/{z}/{x}/{y}.pbf"]
		}
	}
```
`> top`

```
PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                     
```
```
42875 root      20   0  565.8g 228.3g  39.5g S 100.0  91.8   3904:39 tilemaker
```
  
`> free -h`
  ```
                      total        used        free      shared  buff/cache   available
Mem:            248Gi      190Gi      3.2Gi    0.0Ki     55Gi             56Gi
Swap:           0B            0B           0B

  ```

**Reply by systemed** [reply]
Any logging output goes to console (stdout). You can get more logging by running with `--verbose` but I doubt it would be relevant in this case.

There's two possibilities here - either it's slowly deallocating the node/way/relation store which is no longer needed at this stage of the process (`void_mmap_allocator::shutdown();` in tilemaker.cpp), or it's taking a very long time to assemble the low-zoom tiles from the z16 constituents.

Out of interest, what's the reason you're generating tiles at z15/z16?

**Reply by sujithkris** [reply]
> Out of interest, what's the reason you're generating tiles at z15/z16?

As my final plan is to place a building map on this from another source, z14 is too far and the building looks tiny. I may need support to zoom the full map till z20 at least to see the details of the rooms in my building map.

Well, I remember you hinted about reusing z14 pbfs for higher levels.  Currently, for me, `15/x/y.pbf` is requested when the map is zoomed to z15, (and so on) which results in a blank tile as z15 tiles are not present. Whether setting a `maxzoom:14` in style json is sufficient to make the maplibre make requests to `14/x/y.pbf` even at higher zoom levels?

I could see mapbox-gl-js is requesting `16/x/y.pbf` even when the map is zoomed to 20.

**Reply by systemed** [reply]
Yes – if you set maxzoom for that source to 14 in the stylesheet, MBGL won't make requests any higher than that, and it'll just "overzoom" from the data in the z14 tiles. If you set the `"high_resolution": true` parameter in tilemaker's JSON config, then tilemaker will use extra resolution at z14 so the geometries stay sharp when you're zoomed in.

If you're planning on overlaying a building map, I'd suggest you generate the planet just up to z14 with the high_resolution parameter set. Then generate your building map separately.

I might put a warning into tilemaker to say "You really don't need to do this" when people generate tiles above z14/15 because no-one ever understands this about vector tiles ;)

**Comment by dieterdreist** [comment]
sent from a phone

> On 29 Mar 2023, at 13:46, Richard Fairhurst ***@***.***> wrote:
> 
> Yes – if you set maxZoom to 14 in the stylesheet, MBGL won't make requests any higher than that, and it'll just "overzoom" from the data in the z14 tiles. If you set the "high_resolution": true parameter in tilemaker's JSON config, then tilemaker will use extra resolution at z14 so the geometries stay sharp when you're zoomed in.


will it use the original input data resolution or will it still simplify but with a higher resolution?

**Reply by systemed** [reply]
Entirely up to you – you can specify simplification per-layer in tilemaker's JSON config.

**Comment by dieterdreist** [comment]
sent from a phone

> On 29 Mar 2023, at 17:46, Richard Fairhurst ***@***.***> wrote:
> 
> Entirely up to you – you can specify simplification per-layer in tilemaker's JSON config.


yes I know and am using it for one layer, just wanted to know what exactly the high detail flag did (skip simplification or change the threshold)

**Reply by systemed** [reply]
Ah, ok! It increases the vector tile resolution from 4096 to 8192 at the highest zoom level.

**Comment by sujithkris** [comment]
@systemed: 
1. Just wondering about the impact of the `--compact` flag? I could not see any documentation w.r.t that.

2. Whether the coastline/water_polygons.shp is sufficient to address the below difference between openmaptiles+tilemaker and mapbox ?

i) Pier boundaries are not precise, seeing background layer at edges

Mapbox
<img width="683" alt="Screenshot 2023-04-11 at 8 33 43 PM" src="https://user-images.githubusercontent.com/2143097/231205832-b8255163-97df-4247-bb56-c95b2829a767.png">

Tilemaker
<img width="728" alt="Screenshot 2023-04-11 at 8 33 32 PM" src="https://user-images.githubusercontent.com/2143097/231205858-21091219-de1e-4da9-8c19-8c51baa9d683.png">

ii) Ocean Missing in left hand side ( however landlocked water bodies / ocean available)

Mapbox
<img width="499" alt="Screenshot 2023-04-11 at 8 31 47 PM" src="https://user-images.githubusercontent.com/2143097/231206047-7f20f6eb-860d-40b9-a05d-19e76aa58e1e.png">

Tilemaker
<img width="590" alt="Screenshot 2023-04-11 at 8 31 39 PM" src="https://user-images.githubusercontent.com/2143097/231206065-a6c0c5fd-29ad-40ee-8379-13f8b6ffbc96.png">

**Comment by sujithkris** [comment]
@systemed : I generated waterbodies separately without source ( ---input). This time I am getting full water bodies.
But my concern is how I can merge the tiles I generated now with the existing tiles of planet osm pbf. 
I am not generating the single `.mbtiles` file, but rather direct `.pbf` files.

Question: Whether the `--merge` work for `.pbf` files generation? Or limited to `.mbtiles` ?

**Reply by systemed** [reply]
`--merge` is mbtiles-only I'm afraid.

`--compact` is used if you've renumbered your .pbf beforehand (with `osmium renumber`). It saves memory by using contiguous storage for nodes - basically it means tilemaker can use an array/vector rather than a hash/map.

**Comment by FunMiles** [comment]
I found this project from the README of planetiler. I thought (and still think) that it could allow me to play with custom profiles more easily than planetiler allows.
On planetiler, I had no problem doing the whole planet within a few hours on my laptop with 64GB. What is the reason tilemaker needs more memory?

**Reply by systemed** [reply]
Different internal storage basically. In particular tilemaker currently keeps all generated geometries in memory. I'm working on a refactoring in #499 that should enable the memory footprint to be substantially reduced.

**Comment by jopfre** [comment]
if I generate tiles from regions (e.g. asia, europe etc) and then use --merge will this reduce the total ram requirements?
